File: BoxFilter/CPP/src/BoxFilter.cpp
#include <iostream>
#include "BoxFilter.hpp"
#include "numeric"

BoxFilter::BoxFilter(size_t s) : m_deque(s, 0.0)
{
    // nothing needs to be done here
}

void BoxFilter::add(float value)
{
    m_deque.pop_front();
    m_deque.push_back(value);
}

float BoxFilter::update(float value)
{
    add(value);

    if (m_filledSize < m_deque.size())
    {
        ++m_filledSize;
    }

    auto sum = std::accumulate(m_deque.rbegin(),
                               m_deque.rbegin() + m_filledSize, 0.0);

    return sum / m_filledSize;
}

File: BoxFilter/Python/BoxFilter.py
# import modules here
from collections import deque as DQ
from itertools import accumulate

class BoxFilter:
    def __init__(self, size=5):
        self._deque = DQ(maxlen=size)

    def update(self, val):
        self._deque.append(val)
        sum = max(accumulate(self._deque))
        return sum / len(self._deque)


if __name__ == "__main__":
    test_array = [1,3,4,5,6,12,6,15,6,9,22,23,34,43,13,54]
    bf = BoxFilter(5)
    for x in test_array:
        print(bf.update(x))
